nullaway/src/test/java/com/uber/nullaway/NullAwayCoreTests.java

```diff
@@ -990,4 +990,90 @@ public void annotationAppliedToTypeParameter() {
             "}")
         .doTest();
   }
+
+  @Test
+  public void annotationAppliedToInnerTypeImplicitly() {
+    defaultCompilationHelper
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import org.checkerframework.checker.nullness.qual.Nullable;",
+            "class Test {",
+            "  @Nullable Foo f;", // i.e. Test.@Nullable Foo
+            "  class Foo { }",
+            "  public void test() {",
+            "    // BUG: Diagnostic contains: dereferenced",
+            "    f.hashCode();",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void annotationAppliedToInnerTypeExplicitly() {
+    defaultCompilationHelper
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import org.checkerframework.checker.nullness.qual.Nullable;",
+            "class Test {",
+            "  Test.@Nullable Foo f1;",
+            "  @Nullable Test.Foo f2;",
+            "  class Foo { }",
+            "  public void test() {",
+            "    // BUG: Diagnostic contains: dereferenced",
+            "    f1.hashCode();",
+            "    // BUG: Diagnostic contains: dereferenced",
+            "    f2.hashCode();",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void annotationAppliedToInnerTypeExplicitly2() {
+    defaultCompilationHelper
+        .addSourceLines(
+            "Bar.java",
+            "package com.uber;",
+            "import org.checkerframework.checker.nullness.qual.Nullable;",
+            "class Bar {",
+            "  public class Foo { }",
+            "}")
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import org.checkerframework.checker.nullness.qual.Nullable;",
+            "class Test {",
+            "  Bar.@Nullable Foo f1;",
+            "  @Nullable Bar.Foo f2;",
```

Comment:
Here we could potentially filter the first case, by only allowing inner types if the outer types match the containing scope, but it seems like extra brittleness and I am not sure supporting this is incorrect. Is there a difference in semantics here between the two cases above according to JSpecify or anything else?

Reply:
Actually, Error Prone does seem to care about this:

```
error: [NullableOnContainingClass] Type-use nullability annotations should annotate the inner class, not the outer class (e.g., write `A.@Nullable B` instead of `@Nullable A.B`).
```

Which is annoying, because many devs see the annotation as going on the field, not the type, whether or not it is a type use annotation. So they would write:

```
@Nullable
Bar.Foo f2;
```

and get an error from that checker.

Reply:
Still need to look at the whole PR, but yes, according to the spec, type-use annotations should go on the inner type:

https://jspecify.dev/docs/user-guide#type-use-annotation-syntax

We'll have to help developers adjust to this with good error messages.